compiler/modules/front/const.bal

```diff
@@ -517,47 +520,84 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
     }
 }
 
-function foldFloatLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    // This will need to change when we support decimal
-    float|error result = floatFromDecimalLiteral(expr.untypedLiteral);
-    if result is float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+function foldFpLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError { 
+    t:Context tc = cx.typeContext();
+    var { typeSuffix, untypedLiteral, startPos } = expr;
+    float|decimal result;
+    if typeSuffix != () {
+        result = typeSuffix is s:FLOAT_TYPE_SUFFIX ? floatFromFpLiteral(untypedLiteral) : check decimalFromFpLiteral(cx, untypedLiteral, startPos);
+    }
+    else if expectedType == () || expectsFloat(tc, expectedType) {
+        result = floatFromFpLiteral(untypedLiteral);
+    }
+    else if expectsDecimal(tc, expectedType) {
+        result = check decimalFromFpLiteral(cx, untypedLiteral, startPos);
     }
     else {
-        return cx.semanticErr("invalid float literal", cause=result, pos=expr.startPos);
+        result = floatFromFpLiteral(untypedLiteral);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
 function foldIntLiteralExpr(FoldContext cx, t:SemType? expectedType, s:IntLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    float|int|error result;
-    string ty;
-    if expr.base == 10 && expectsFloat(expectedType) {
-        result = floatFromDecimalLiteral(expr.digits);
-        ty = "float"; 
-    }
-    else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
+    t:Context tc = cx.typeContext();
+    Position startPos = expr.startPos;
+    int|float|decimal result;
+    if expectedType == () || expectsInt(tc, expectedType) {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if expectsFloat(tc, expectedType) {
+        result = floatFromFpLiteral(expr.digits);
     }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+    else if expectsDecimal(tc, expectedType) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
+}
+
+function expectsInt(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:INT));
```

Comment:
Better to use `t:isNever` here.  Also we have a lot of cases where we are doing `t:isNever(t:intersect(semType, bitSet))`. So we should have a function in types that does it directly (by looking at some|all in ComplexSemType case) and get rid of these three functions. What to call it? I suggest `t:includesSome`. So instead of `expectsFloat(tc, expectedType)` we have `t:includesSome(expectedType, t:FLOAT)`

Reply:
Fixed